### DISCUSSION TITLE
Modify sqlalchemy query

### DISCUSSION THREAD

**Discussion by MAM-SYS** [sqlalchemy/sqlalchemy#7928] [Usage Questions] [answered]
Hello dear friends

In our company we built an OData Parser that convert OData to sqlalchemy query.
But in very rare cases this parser returns some wrong queries. (Like with an unneeded duplicated join on same model)
My question is whether there is a way to modify or edit the returned query? like removing the unneeded duplicated join
As we can add a join, is there any way to remove it?

Thanks a lot

**Comment by CaselIT** [comment]
Hi,

I don't thins if there is any api to do that at the moment, but maybe the [traversals module](https://docs.sqlalchemy.org/en/14/core/visitors.html) could be used for this? 

@zzzeek do you have suggestions?

**Reply by MAM-SYS** [reply]
I appreciate a lot for your help. I'll check this module

**Comment by zzzeek** [comment, marked as answer]
we don't really have public API for removing a join.   the "traversals" aspect of it is not necessarily needed, you would just be altering the `_from_obj` collection of the query object.

Overall I would recommend that you are using `select()` constructs and not ORM `Query` objects and that you are on SQLAlchemy 1.4.  

Also, instead of removing a redundant join, I would modify your parser to check the existing FROMs of the SELECT **before** adding the join that's redundant.    for more intricate cases, that's where the traversals module might be more useful.   this still involves semi-public APIs (looking inside of underscore collections on select()).

**Reply by MAM-SYS** [reply]
Thanks a lot @zzzeek 
I consider all these recommendations
